### PR TITLE
add xclbin URI support

### PR DESCRIPTION
**Pull request by timsnyder-siv**
use fsspec to enable xclbin's to be one of any URI protocol
supported by the library or an installed add-on

#### UI / API Impact

Enables xclbin in hwdb to be a URI supported by [fsspec](https://filesystem-spec.readthedocs.io/en/latest/index.html)

#### Verilog / AGFI Compatibility

No impact to Verilog or AGFI.

### Contributor Checklist
- [X] Is this PR's title suitable for inclusion in the changelog and have you added a `changelog:<topic>` label?
- [x] Did you add Scaladoc/docstring/doxygen to every public function/method?
- [x] Update the docs with 
  - list of protocols tested and known to work
  - pointer to fsspec config that can be used to inject auth stuff (TO ALL CALLS).  Since the auth args go to all calls when using the fsspec level config, it won't work in all cases but it might be useful as we start to use things and determine if/how we want to add features to the manager to specify auth.  In the limit, you might need different auth for every file you interact with but you don't want to have to repeat and maintain auth everywhere.  Just brainstorming, it seems like fsspec should allow it's config to do URI pattern matching to provide auth.
- [x] Did you add at least one test demonstrating the PR?
- [X] Add a test for `ssh://` in #1167 
- [X] Did you delete any extraneous prints/debugging code?
- [X] Did you state the UI / API impact?
- [X] Did you specify the Verilog / AGFI compatibility impact?

- [x] If applicable, did you regenerate and publicly share default AGFIs?

- [X] If applicable, did you apply the `ci:fpga-deploy` label?

- [X] If applicable, did you apply the `Please Backport` label?

### Reviewer Checklist (only modified by reviewer)
- [ ] Is the title suitable for inclusion in the changelog and does the PR have a `changelog:<topic>` label?
- [ ] Did you mark the proper release milestone?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?

### CODE REVIEW

File: deploy/runtools/runtime_config.py
Comment (abejgonzalez) [inline deploy/runtools/runtime_config.py:43]: IMO we should just enforce that this is a URI (i.e. enforce `file:///path/to/file`)

File: docs/Advanced-Usage/Manager/Manager-Configuration-Files.rst
Comment (abejgonzalez) [inline docs/Advanced-Usage/Manager/Manager-Configuration-Files.rst:506]: Ditto.

File: deploy/runtools/run_farm_deploy_managers.py
Comment (abejgonzalez) [inline deploy/runtools/run_farm_deploy_managers.py:14]: Shouldn't we add this to the `machine-launch-script.sh` even if it is pulled in from something else?

File: deploy/runtools/run_farm_deploy_managers.py
Comment (abejgonzalez) [inline deploy/runtools/run_farm_deploy_managers.py:674]:
```suggestion
            def get_xclbin(xclbin_uri: str, xclbin_lpath: PathLike) -> None:
```

File: deploy/runtools/run_farm_deploy_managers.py
Comment (abejgonzalez) [inline deploy/runtools/run_farm_deploy_managers.py:680]: This code should run in parallel on multiple run hosts but once you are using a specific run host this is done serially: 

https://github.com/firesim/firesim/blob/7839d4fda5eb360841694833c40f50e3e9bf3387/deploy/runtools/firesim_topology_with_passes.py#L388-L393

https://github.com/firesim/firesim/blob/7839d4fda5eb360841694833c40f50e3e9bf3387/deploy/runtools/run_farm_deploy_managers.py#L598

Comment (abejgonzalez) [reply deploy/runtools/run_farm_deploy_managers.py:680]: I think using the file cache should be fine (and doesn't need a lock)

File: deploy/runtools/run_farm_deploy_managers.py
Comment (abejgonzalez) [inline deploy/runtools/run_farm_deploy_managers.py:683]:
```suggestion
                rootLogger.debug(f"Downloading {xclbin_uri} to {lpath}")
```

File: deploy/runtools/run_farm_deploy_managers.py
Comment (abejgonzalez) [inline deploy/runtools/run_farm_deploy_managers.py:681]: I hate to say this... but we should probably port to using `pathlib` everywhere. (just a comment)

File: deploy/runtools/run_farm_deploy_managers.py
Comment (abejgonzalez) [inline deploy/runtools/run_farm_deploy_managers.py:685]: Whats the debug output if you can't obtain a file? Does this get caught by the `StreamLogger` (I forget how `StreamLogger` works exactly)

**Review by abejgonzalez — CHANGES_REQUESTED** [review]
Seems pretty good to me so far minus some small nits.

File: deploy/runtools/run_farm_deploy_managers.py
Comment (timsnyder-siv) [reply deploy/runtools/run_farm_deploy_managers.py:14]: Oh, yeah, if I didn't add fsspec to machine-launch-script, that was an oversight.  I'll add it.

File: deploy/runtools/run_farm_deploy_managers.py
Comment (timsnyder-siv) [reply deploy/runtools/run_farm_deploy_managers.py:680]: Ok.  It's good to know that it should work now.  What happens if someone changes the code so that a lock is needed?  How do we detect that case and blow up?

Said differently, can we test to ensure that it should continue to be okay?

File: deploy/runtools/run_farm_deploy_managers.py
Comment (timsnyder-siv) [reply deploy/runtools/run_farm_deploy_managers.py:685]: Good question.  I think it throws an exception and I would suspect that the debug output quality varies based on which filesystem implementation is being used.  When I tried to `fs.ls()` a non-existent file (outside of FireSim and fabric), a stacktrace looked like :
```
FileNotFoundError                         Traceback (most recent call last)
/home/centos/orca_pr_220621/fsspec.ipynb Cell 13 in <cell line: 1>()
----> [1](vscode-notebook-cell://ssh-remote%2Bfs5/home/centos/orca_pr_220621/fsspec.ipynb#ch0000012vscode-remote?line=0) fs[0].ls(fs[1])

File ~/miniforge3/envs/jlab/lib/python3.10/site-packages/fsspec/implementations/sftp.py:115, in SFTPFileSystem.ls(self, path, detail)
    113 def ls(self, path, detail=False):
    114     logger.debug("Listing folder %s" % path)
--> 115     stats = [self._decode_stat(stat, path) for stat in self.ftp.listdir_iter(path)]
    116     if detail:
    117         return stats

File ~/miniforge3/envs/jlab/lib/python3.10/site-packages/fsspec/implementations/sftp.py:115, in <listcomp>(.0)
    113 def ls(self, path, detail=False):
    114     logger.debug("Listing folder %s" % path)
--> 115     stats = [self._decode_stat(stat, path) for stat in self.ftp.listdir_iter(path)]
    116     if detail:
    117         return stats

File ~/miniforge3/envs/jlab/lib/python3.10/site-packages/paramiko/sftp_client.py:278, in SFTPClient.listdir_iter(self, path, read_aheads)
    276 path = self._adjust_cwd(path)
    277 self._log(DEBUG, "listdir({!r})".format(path))
--> 278 t, msg = self._request(CMD_OPENDIR, path)
    280 if t != CMD_HANDLE:
    281     raise SFTPError("Expected handle")

File ~/miniforge3/envs/jlab/lib/python3.10/site-packages/paramiko/sftp_client.py:822, in SFTPClient._request(self, t, *arg)
    820 def _request(self, t, *arg):
    821     num = self._async_request(type(None), t, *arg)
--> 822     return self._read_response(num)

File ~/miniforge3/envs/jlab/lib/python3.10/site-packages/paramiko/sftp_client.py:874, in SFTPClient._read_response(self, waitfor)
    871 if num == waitfor:
    872     # synchronous
    873     if t == CMD_STATUS:
--> 874         self._convert_status(msg)
    875     return t, msg
    877 # can not rewrite this to deal with E721, either as a None check
    878 # nor as not an instance of None or NoneType

File ~/miniforge3/envs/jlab/lib/python3.10/site-packages/paramiko/sftp_client.py:903, in SFTPClient._convert_status(self, msg)
    900     raise EOFError(text)
    901 elif code == SFTP_NO_SUCH_FILE:
    902     # clever idea from john a. meinel: map the error codes to errno
--> 903     raise IOError(errno.ENOENT, text)
    904 elif code == SFTP_PERMISSION_DENIED:
    905     raise IOError(errno.EACCES, text)

FileNotFoundError: [Errno 2] No such file
```

I suspect that we will want to put this into  a `try/except` block that makes the debug output nicer than that big stacktrace.

File: deploy/runtools/runtime_config.py
Comment (timsnyder-siv) [reply deploy/runtools/runtime_config.py:43]: Probably.  The only thing we would be losing is the ability for someone to put a relative path in their hwdb (because file URIs need the full path).

While on this topic, I wonder how fsspec would handle `file://<ipaddress>/path/to/file` because that is a valid file URI.  https://en.wikipedia.org/wiki/File_URI_scheme. I had forgotten that you can give a hostname in a file URI because I never do.

Comment (abejgonzalez) [reply deploy/runtools/runtime_config.py:43]: I don't think that having a relative path is that important. I think everything using a URI would be clearer to users.

As for having a hostname in a URI, I assume it probably just does a `rsync` like operation to retrieve the file. Whats your concern?

File: deploy/runtools/run_farm_deploy_managers.py
Comment (abejgonzalez) [reply deploy/runtools/run_farm_deploy_managers.py:680]: I vote to just have a comment indicating that the code should not be run in parallel on the same host. I think because F1 requires that we do things serially, we will just enforce this in the future.

File: deploy/runtools/run_farm_deploy_managers.py
Comment (davidbiancolin) [inline deploy/runtools/run_farm_deploy_managers.py:666]: If you could provide a helpful assertion message, that might be useful.

File: deploy/runtools/run_farm_deploy_managers.py
Comment (davidbiancolin) [inline deploy/runtools/run_farm_deploy_managers.py:682]: This is going to fail often since people rerun infrasetup before runworkload right? For all other files, the current behavior of infrasetup is to silently clobber stuff that exists so we should probably do the same here no?

However, we could have the manager backup the directory on re-invocations of infrasetup or ensure the directory does not exist, because the behavior here is probably more desirable.

File: deploy/runtools/runtime_config.py
Comment (davidbiancolin) [reply deploy/runtools/runtime_config.py:43]: authentication will be annoying depending on how it's implemented under the hood. 

I personally would like relative paths to be supported.

File: deploy/runtools/run_farm_deploy_managers.py
Comment (timsnyder-siv) [reply deploy/runtools/run_farm_deploy_managers.py:30]: I swiped it from https://github.com/pandas-dev/pandas/blob/96b036cbcf7db5d3ba875aac28c4f6a678214bfb/pandas/io/common.py#L73. I'll add a comment that points there unless you were saying that you'd like a pointer to https://datatracker.ietf.org/doc/html/rfc3986

File: deploy/runtools/run_farm_deploy_managers.py
Comment (timsnyder-siv) [reply deploy/runtools/run_farm_deploy_managers.py:666]: Sure.  I'll come up with something and add it to the other methods that make this same assertion without a message.

File: deploy/runtools/run_farm_deploy_managers.py
Comment (timsnyder-siv) [reply deploy/runtools/run_farm_deploy_managers.py:682]: Yeah.  You make a good point about this failing often.  The ideal thing to do would be using the FileCache so that fsspec will do a checksum and not re-download the file if you've already got it.  I'd just like to be confident that it is safe to use the FileCache (and that it remains safe).  

I really like the current behavior of infrasetup where the rsync is essentially a no-op if you need to rerun it.  I wouldn't want to change that (although I'm not sure that's what you were suggesting with your 'backup' idea).

File: deploy/runtools/runtime_config.py
Comment (timsnyder-siv) [reply deploy/runtools/runtime_config.py:43]: > Whats your concern?

That we would need to document it because users will need to know how to setup their authentication.

File: deploy/runtools/run_farm_deploy_managers.py
Comment (abejgonzalez) [reply deploy/runtools/run_farm_deploy_managers.py:682]: If you are concerned about this potentially breaking... I think grabbing a lock and releasing it should be fine (esp. since that wouldn't be the bottleneck).

Comment (timsnyder-siv) [reply deploy/runtools/run_farm_deploy_managers.py:682]: @abejgonzalez what kind of lock are you thinking?  `flock()`?  I typically stay away from `flock()` because it isn't implemented well in many of the NFSv3 environments I find myself having to work in.  I think I'll just change the code to overwrite the file for now and if we can revisit the FileCache in a future PR if needed.

**Comment by timsnyder-siv** [comment]
I have some more commits that I want to push to this after rebasing on top of the fixed pytests (because they weren't really broken by this stuff)

File: deploy/runtools/runtime_config.py
Comment (abejgonzalez) [reply deploy/runtools/runtime_config.py:43]: If relative paths are supported, I think it would be nice to enforce that full paths are done using URIs while relative paths shouldn't (maybe `pathlib` has a way to enforce a `Path` is a relative path)? However, this isn't blocking so if that can't happen / is too much effort we can just do what currently exists.

File: deploy/runtools/run_farm_deploy_managers.py
Comment (abejgonzalez) [reply deploy/runtools/run_farm_deploy_managers.py:685]: I think this debug output is fine. I rather it be very verbose and descriptive without us having to do extra work to catch an exception.

File: deploy/runtools/run_farm_deploy_managers.py
Comment (abejgonzalez) [reply deploy/runtools/run_farm_deploy_managers.py:666]: I think this assertion should just be put into a function since this is done quite a bit.

File: deploy/runtools/run_farm_deploy_managers.py
Comment (abejgonzalez) [reply deploy/runtools/run_farm_deploy_managers.py:682]: SGTM. I didn't have a lock in mind.

File: deploy/runtools/run_farm_deploy_managers.py
Comment (timsnyder-siv) [inline deploy/runtools/run_farm_deploy_managers.py:14]: Looks like fsspec doesn't have type annotations or hints available yet but they aren't opposed to discussing them if this becomes problematic. https://github.com/fsspec/filesystem_spec/issues/625

File: deploy/runtools/run_farm_deploy_managers.py
Comment (timsnyder-siv) [inline deploy/runtools/run_farm_deploy_managers.py:30]:
```suggestion
# from  https://github.com/pandas-dev/pandas/blob/96b036cbcf7db5d3ba875aac28c4f6a678214bfb/pandas/io/common.py#L73
_RFC_3986_PATTERN = re.compile(r"^[A-Za-z][A-Za-z0-9+\-+.]*://")
```

@davidbiancolin is this the kind of ref you are wanting?

File: deploy/runtools/runtime_config.py
Comment (timsnyder-siv) [reply deploy/runtools/runtime_config.py:43]: It wouldn't be hard to `raise` something if `localize_xclbin()` is given an absolute filepath.  I'm concerned that it would be better for this to be caught earlier like when the config is read and validated against a schema.

File: deploy/runtools/run_farm_deploy_managers.py
Comment (timsnyder-siv) [reply deploy/runtools/run_farm_deploy_managers.py]: added in https://github.com/firesim/firesim/pull/1140/commits/7e93452b96b7d2aa26d5c2ee5483303d580ccfec

File: deploy/runtools/run_farm_deploy_managers.py
Comment (timsnyder-siv) [reply deploy/runtools/run_farm_deploy_managers.py]: done. see https://github.com/firesim/firesim/pull/1140/commits/da27c75ed25b328b4cf7c68baf919f000fd37613

File: deploy/runtools/run_farm_deploy_managers.py
Comment (timsnyder-siv) [reply deploy/runtools/run_farm_deploy_managers.py]: done. see https://github.com/firesim/firesim/pull/1140/commits/da27c75ed25b328b4cf7c68baf919f000fd37613

File: deploy/runtools/run_farm_deploy_managers.py
Comment (timsnyder-siv) [reply deploy/runtools/run_farm_deploy_managers.py]: done. see https://github.com/firesim/firesim/pull/1140/commits/da27c75ed25b328b4cf7c68baf919f000fd37613

File: deploy/runtools/run_farm_deploy_managers.py
Comment (timsnyder-siv) [reply deploy/runtools/run_farm_deploy_managers.py]: cool.  I won't worry about adding nicer error messaging for now then.

File: deploy/runtools/run_farm_deploy_managers.py
Comment (timsnyder-siv) [reply deploy/runtools/run_farm_deploy_managers.py]: Yeah, I'm going to punt on using the FileCache for now.  I made `localize_xclbin()` overwrite with a logged message in https://github.com/firesim/firesim/pull/1140/commits/9bd9bce1d0880909a3676b967d3f967dfd368bfd

File: docs/Advanced-Usage/Manager/Manager-Configuration-Files.rst
Comment (timsnyder-siv) [inline docs/Advanced-Usage/Manager/Manager-Configuration-Files.rst:524]: Should tweak this to explain the fsspec config file mechanism for injecting `kwargs` for backends that require them.  Upstream doc is at https://filesystem-spec.readthedocs.io/en/latest/features.html#configuration

File: deploy/runtools/run_farm_deploy_managers.py
Comment (timsnyder-siv) [reply deploy/runtools/run_farm_deploy_managers.py:682]: Yeah, it might make sense.  I've started using it only because it makes `Path` methods available to tab-completion in whatever IDE I'm using and I'm too lazy to remember which functions need to be imported from `os.path` to do whatever I'm doing.

Note that there are some significant wrinkles with regards to `pathlib` and `fsspec` interoperability (see https://github.com/fsspec/filesystem_spec/issues/539).  Enough so that folks at QuantSight have developed https://github.com/fsspec/universal_pathlib as an adapter for `pathlib` usage with `fsspec`

Comment (abejgonzalez) [reply deploy/runtools/run_farm_deploy_managers.py:682]: Should we use `universal_pathlib` instead of `pathlib` everywhere then? Based on this it seems like it's a superset of functionality? https://github.com/fsspec/universal_pathlib/blob/main/notebooks/examples.ipynb.

I'm slightly concerned about future bugs when using `Path`s (or `UPath`s) with Fabric since `Path` objects are relative to localhost (i.e. `Path("/remote/path").mkdir()` will create a local directory when you meant to do this with Fabric). In any case, this discussion should happen in another PR.

Comment (timsnyder-siv) [reply deploy/runtools/run_farm_deploy_managers.py:682]: I agree that we should pull this out into another PR.  I captured #1148 for discussion.

File: deploy/runtools/run_farm_deploy_managers.py
Comment (timsnyder-siv) [reply deploy/runtools/run_farm_deploy_managers.py]: Or removed because letting Python raise IndexError is probably just as useful and reduces the amount of code in the manager.  I'll leave that to future cleanup PRs .

Comment (timsnyder-siv) [reply deploy/runtools/run_farm_deploy_managers.py]: Also, if we move it into another function, I wonder if mypy is smart enough to use the assert in it's type-checking.  I know this particular assert predates the type checking and this is more a curiosity that I'm voicing out-loud because we have quite a few mypy helper asserts that check for `None`ness of `Optional` types and I could envision asking the same thing about them.

File: deploy/runtools/run_farm_deploy_managers.py
Comment (davidbiancolin) [reply deploy/runtools/run_farm_deploy_managers.py:30]: Perfect.

**Comment by timsnyder-siv** [comment]
I updated the checklist with the remaining tasks for this PR to be ready to merge.

File: deploy/tests/test_utils.py
Comment (timsnyder-siv) [inline deploy/tests/test_utils.py:17]: Do we want to add a test for ssh?  The simplest way to do it would be to spin up a docker container that the test would try to pull things out of with ssh.  If docker isn't installed, the test would be skipped.

Alternately, we write a test for this that utilizes the multiple hosts in CI.  Or we could do both so that devs can test this locally (if they have docker installed) and we have a test in CI without having to install docker in the place where we're running pytests.

File: deploy/tests/test_utils.py
Comment (filipstamenkovic-sifive) [inline deploy/tests/test_utils.py:32]:
```suggestion
            mock_s3_client = boto3.client('s3', region_name='us-west-2')
            mock_s3_client.create_bucket(
                Bucket="TestBucket",
                CreateBucketConfiguration={
                   'LocationConstraint': 'us-west-2',
                }
            )
```

File: scripts/machine-launch-script.sh
Comment (filipstamenkovic-sifive) [inline scripts/machine-launch-script.sh:311]:
```suggestion
        fsspec \
        s3fs \
```

**Review by filipstamenkovic-sifive — COMMENTED** [review]
@timsnyder-siv adding a few suggestions to fix the test issue in CI.

File: deploy/util/io.py
Comment (abejgonzalez) [inline deploy/util/io.py:28]:
```suggestion
```

File: deploy/runtools/run_farm_deploy_managers.py
Comment (abejgonzalez) [reply deploy/runtools/run_farm_deploy_managers.py]: Mypy is smart enough to do this. I can do quick cleanup of this when I get the chance

File: deploy/tests/test_utils.py
Comment (abejgonzalez) [reply deploy/tests/test_utils.py:17]: Couldn't we just test this by running `localhost` for the "it should work" case and some invalid IP for the "it shouldn't work" case? I'm not sure it's worth having docker installed for such a small test

File: docs/Advanced-Usage/Manager/Manager-Configuration-Files.rst
Comment (timsnyder-siv) [reply docs/Advanced-Usage/Manager/Manager-Configuration-Files.rst]: Done in https://github.com/firesim/firesim/pull/1140/commits/2268d442ee4d36e4f74f7487c5a45316d8169b5f

File: deploy/runtools/run_farm_deploy_managers.py
Comment (timsnyder-siv) [reply deploy/runtools/run_farm_deploy_managers.py]: We decided to punt on FileCache for this PR.

File: deploy/tests/test_utils.py
Comment (timsnyder-siv) [reply deploy/tests/test_utils.py:17]: @abejgonzalez how do we differentiate whether someone doesn't have sshd configured correctly on localhost vs the test actually working?  I don't like having a dependency into sshd that the tests aren't actually controling the configuration for.

A lighter-weight option could be for a test fixture to spin up a Paramiko sshd on a non-system port or use something like [this gist](https://gist.github.com/vxgmichel/d0b7e4cc3caab32601051ee262ee7b31).   What do you think?

File: deploy/runtools/runtime_config.py
Comment (timsnyder-siv) [reply deploy/runtools/runtime_config.py:42]: @abejgonzalez, the more I think about this, the less I feel like enforcing where users use a `file://` URI vs a simple string path.  Please speak up if you feel strongly enough about this that it should block merging.

Comment (timsnyder-siv) [reply deploy/runtools/runtime_config.py:42]: @abejgonzalez , since you requested changes, I think I need you to approve.  I guess there are new mypy errors I need to cleanup also.

File: deploy/tests/test_utils.py
Comment (abejgonzalez) [reply deploy/tests/test_utils.py:17]: That seems better to me.

**Review by abejgonzalez — APPROVED** [review]
LGTM. Pending all CI passing.

File: deploy/tests/test_utils.py
Comment (timsnyder-siv) [reply deploy/tests/test_utils.py:17]: I've asked @filipstamenkovic-sifive to help me add a test.

Comment (timsnyder-siv) [reply deploy/tests/test_utils.py:17]: Captured in #1167  for followup.  I'd like to go ahead and get this merged if the CI passes this time.

File: deploy/runtools/run_farm_deploy_managers.py
Comment (sifive-benjamin-morse) [inline deploy/runtools/run_farm_deploy_managers.py:655]: From my testing, this line doesn't compile. run() only accepts a string, which is a shell command, not a function. The 2nd and 3rd arguments are booleans, not arguments to the first function. Am I on the wrong version of fabric or something?  https://github.com/fabric/fabric/blob/1.14/fabric/operations.py#L970

Comment (timsnyder-siv) [reply deploy/runtools/run_farm_deploy_managers.py:655]: I think that we meant `execute()` instead of `run()`.

**Pull request merged**
